### ISSUE TITLE
new lib: Tween (for easy animations)

### ISSUE THREAD

**Issue by achimmihca** [UltraStar-Deluxe/Play#50]
There are some easy to use libs for Unity that make animating GameObjects  (e.g. move to XY, or fade color to XXX) a lot easier. I want to use one of these for USDP. What I found so far:

**iTween**: [assetstore](https://assetstore.unity.com/packages/tools/animation/itween-84), [github](https://github.com/jtothebell/iTween/blob/master/iTween.cs)
It is a popular one, I think one of the first (if not THE first) Tween lib for Unity and I used it in another project before.
It is a ~8000 LoC single file. However, it uses an API with key-value maps that are not type safe, thus relatively brittle, and of course without code completion of an IDE. For this reason I don't want to use it.
Furthermore, LeanTween (see below) did a nice performance comparision and iTween seems to be quite slow compared to the rest.

**LeanTween**: [assetstore](https://assetstore.unity.com/packages/tools/animation/leantween-3595), [github](https://github.com/dentedpixel/LeanTween/)
This looks very good. Simple to use and able to do a lot of stuff out of the box. I tried the examples and you can do simple stuff with understandable and type safe one-liners. More complex stuff is composed of the simple ones using sequences.
Also all relevant platforms are supported. Very good performance compared to other Tween libs (see diagram in the assetstore images).
Also nice [documentation](http://dentedpixel.com/LeanTweenDocumentation/classes/LeanTween.html). The lib is available as source code or compiled as DLL, such that the source files do not re-compile all the time in the Unity Editor.
License is MIT, thus a perfect match for USDP.

**DOTween**: [assetstore](https://assetstore.unity.com/packages/tools/animation/dotween-hotween-v2-27676), [github](https://github.com/Demigiant/dotween)
This looks like another good choice, similar to LeanTween. However, there is a pro version, which is paid. For an open project like UltraStar Deluxe Play I think this is no go because it implies that some features are missing in the free version and future development is not in community hands.
Furthermore, LeanTween seemed to perform a bit better.

**Tween**: [assetstore](https://assetstore.unity.com/packages/tools/animation/tween-55983)
This is a very simple script. Basically the core concepts of a tween system (i.e. interpolating values for floats, Vectors, Colors) is implemented in ~1000 lines (including comments) in a single file. License is MIT, thus a perfect match for USDP.
This is super easy and what it does, it does well.
However, it does not have all the features of LeanTween or DOTween, e.g. creating sequences of animations, delays and looping will require some manual work.
I still think it should do for USDP, at least for now.

What do you think?
In conclusion, to me LeanTween seems to be the best choice for USDP. It just looks like it gets the job done with little effort and nice out-of-the-box features. No matter if we need simple or more complex animations. These features of course come at the cost of a bigger lib than the simple Tween lib, but I think this is not a problem.
I would give LeanTween a try.

**Comment by basisbit** [comment]
**DOTween**: their license states `You can modify the code only for your own use and you cannot redistribute modified versions.` - that doesn't fit well with a MIT licensed project. Would prefer to rather not use it, if avoidable without too much work.
**iTween**: no type-safety - could probably be changed without too much effort to include enum-based type-validation. Also, the github link you mentioned is not from the iTween dev and very old, Can't find any official current repo on github which includes iTween or has a license file.
**Tween**: looks good, although looking at the code, I wonder what benefit this code would bring. Afaik you can do these things already with Unity 5+ API like for example [AnimationCurve](https://docs.unity3d.com/Manual/animeditor-AnimationCurves.html) and this would be unity standard functions, thus making it easier for others to modify/contribute. Anyways, license is ok, code looks ok and small enough to be maintained by us if necessary, so if you want to, feel free to use it.
**LeanTween**: looks like a perfect match, except that the developer suddenly disappeared from the internet about one year ago and does not reply to anything anymore on facebook, twitter and some support forums. Anyways, license is ok, code looks ok, so if you want to, feel free to use it.

Could you maybe try using the current built-in Unity animation features for a bit, and then decide on what would be best as you see fit?

**Comment by achimmihca** [comment]
I just had a look at AnimationCurve. This seems to be a powerful tool that can do complex stuff. However, doing simple stuff with little effort seems not that easy - at least not with out-of-the-box one-liners and easy composition of animations. The AnimationCurve seems to be a perfect tool for animators who work with the Unity Editor rather than by code.

To get an idea of working with AnimationCurve programatically [see here](https://answers.unity.com/questions/911169/procudurally-generate-animationclip-at-runtime.html).
You have to (1) create an AnimationClip, (2) setup the AnimationCurve including setting keyframes, (3) add the curves to the clip for a specific property. Thereby the property is given as a string and thus not type-safe.

Furthermore, LeanTween also comes with handy pre-defined ease types that lets you experiment with how interpolation is applied, e.g. linearly, start slow and finish fast, overshoot and bounce back, ...

So in conclusion, I think LeanTween is the better tool for easy but good looking, coded animations to make things feel more alive and entertaining.

**Comment by achimmihca** [comment]
LeanTween is now used in the part of the master. I added a comment in [Compiling the game ](https://github.com/UltraStar-Deluxe/Play/wiki/Compiling-the-game) how to download the dependencies.

I close this issue now.
